generic/src/main/scala-3/neotypes/generic/deriveProductMap.scala

```diff
@@ -0,0 +1,41 @@
+package neotypes
+package generic
+
+import neotypes.mappers.ResultMapper
+import neotypes.mappers.ResultMapper.DerivedProductMap
+import neotypes.model.types.NeoObject
+import neotypes.model.exceptions.ResultMapperException
+import neotypes.model.exceptions.IncoercibleException
+
+import shapeless3.deriving.*
+
+trait CaseClassDerivedProductMap[P <: Product] extends DerivedProductMap[P]
+object CaseClassDerivedProductMap:
+  private type Acc = (Int, NeoObject, Option[ResultMapperException])
+  given [P <: Product](using
+    labelled: Labelling[P],
+    inst: K0.ProductInstances[ResultMapper, P]
+  ): CaseClassDerivedProductMap[P] with
+    override def map(obj: NeoObject): Either[ResultMapperException, P] =
+      // `unfold` traverses over Product field types.
+      // For example, `inst.unfold` for `case class A(i:Int, s: String)`
+      // traverses over `Int` and `String`
+      val ((_, _, decodeOpt), maybeDerived) = inst.unfold[Acc]((0, obj, None)):
+        [t] =>
+          (acc: Acc, mapper: ResultMapper[t]) =>
+            val (idx, remaining, optErr) = acc
+            val fieldName = labelled.elemLabels(idx)
+            val decodedHead = obj.getAs(fieldName, mapper)
+            (
+              (idx + 1, obj, decodedHead.left.toOption),
```

Comment:
I don't understand what is `optErr`, why do we get the `left` here?

Reply:
some variable names are renamed and the control structure is a bit rewritten for clarity.

Reply:
The result of unfold is either a product type without error or the very first decode error and other cases should not happen.

https://github.com/neotypes/neotypes/pull/667/files#diff-07e5571fdd32b0141267ce8f7eb5a3bd0d726901fbe97b5b127b54277fe767deR41-R54

Reply:
solved by another approach https://github.com/neotypes/neotypes/pull/667#discussion_r1310603282